rogg_granule.c

```diff
@@ -0,0 +1,160 @@
+   LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+   NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+   SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+/* ogg vorbis logical stream granule adjustment using the rogg library */
+
+/* compile with
+   gcc -O2 -g -Wall -I. -o rogg_granule rogg.c rogg_granule.c
+*/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/mman.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <rogg.h>
+
+int granule_adjust = 0;
+
+void print_usage(FILE *out, char *name)
+{
+  fprintf(stderr, "Script for editing ogg headers\n");
+  fprintf(stderr, "%s [-g n] <file1.ogg> [<file2.ogg>...]\n",
+	name);
+  fprintf(stderr,
+		  "    -g n        change the granule position of a logical vorbis stream by n\n"
+		  "                (can be positive or negative)\n"
+		  "\n");
+}
+
+void parse_args(int *argc, char *argv[])
+{
+  int arg = 1;
+  int shift;
+
+  while (arg < *argc) {
+    shift = 0;
+    if (argv[arg][0] == '-') {
+      switch (argv[arg][1]) {
+	case 'g':
+	  /* read granule adjustment from the next arg */
+	  if (sscanf(argv[arg+1], "%i", &granule_adjust) != 1) {
+	    fprintf(stderr, "Could not parse granule adjustment '%s'.\n", argv[arg+1]);
+	    granule_adjust = 0;
+	  } else {
+	    fprintf(stdout, "Adjusting granule position by %i\n", granule_adjust);
+	  }
+	  shift = 2;
+	  break;
+      }
+    }
+    if (shift) {
+      int left = *argc - arg - shift;
+      if (left < 0) {
+	fprintf(stderr, "Internal error parsing argument '%s'.\n", argv[arg]);
+	exit(1);
+      }
+      memmove(&argv[arg], &argv[arg+shift], left*sizeof(*argv));
+      *argc -= shift;
+    } else {
+      arg++;
+    }
+  }
+}
+
+int main(int argc, char *argv[])
+{
+  int f, i;
+  unsigned char *p, *q, *o, *e;
+  struct stat s;
+  rogg_page_header header;
+  uint64_t granulepos;
+
+  parse_args(&argc, argv);
+  if (argc < 2) {
+    print_usage(stderr, argv[0]);
+    exit(1);
+  }
+
+  for (i = 1; i < argc; i++) {
+    f = open(argv[i], O_RDWR);
+    if (f < 0) {
+	fprintf(stderr, "couldn't open '%s'\n", argv[i]);
+	continue;
+    }
+    if (fstat(f, &s) < 0) {
+	fprintf(stderr, "couldn't stat '%s'\n", argv[i]);
+	close(f);
+	continue;
+    }
+    p = mmap(0, s.st_size, PROT_READ|PROT_WRITE,
+	MAP_SHARED, f, 0);
+    if (p == MAP_FAILED) {
+	fprintf(stderr, "couldn't mmap '%s'\n", argv[i]);
+	close(f);
+	continue;
+    }
+    fprintf(stdout, "Checking Ogg file '%s'\n", argv[i]);
+    e = p + s.st_size; /* pointer to the end of the file */
+    q = rogg_scan(p, s.st_size); /* scan for an Ogg page */
+    if (q == NULL) {
+	fprintf(stdout, "couldn't find ogg data!\n");
+    } else {
+      if (q > p) {
+	fprintf(stdout, "Skipped %d garbage bytes at the start\n", (int)(q-p));
+      }
+      while (q < e) {
+	o = rogg_scan(q, e-q); /* find the next Ogg page */
```

Comment:
It would be more robust here to just skip all the header pages. They will all have a zero granulepos (assuming the file is valid) and so the logic below will skip offsetting the value, but it's possible that the first data page *also* has a zero granulepos. In that case it would be better to adjust it, but not the header pages.